### PR TITLE
mysql: more robust release of ephemeral buf

### PR DESCRIPTION
**Pull request by sougou**
@alainjobart @slanning This is a trial fix for the recent vtexplain crashes.

Basically, I found some error code paths that can cause the ephemeral buffer to not be released, which later leads to weirdness. This is a test fix to see if it fixes the vtexplain problems.

Can you test this branch and see what you get?

### CODE REVIEW

**Comment by alainjobart** [comment]
The only errors that I see could happen are packing errors into the buffer. And if they do, then the state is not cleared, and the next time we try to use the buffer, it will trigger an error.

In short, I am not sure this is fixing anything...

(Note I don't have a better explanation, so I don't mind trying this, and I kinda like the clearer start / defer end, even though it's more code).

**Comment by slanning** [comment]
I changed this locally:

    if c.currentEphemeralBuffer == nil {
        // panic("c.currentEphemeralBuffer is nil")
        log.Warningf("c.currentEphemeralBuffer is nil")
    }

because everything panicked immediately. I'm not sure if that should actually not be nil. (?) That warning then appears only when an error happens.

I'm running my query.log replay script, waiting for it to crash. Sometimes it can take hours, even replaying a query.log that has previously caused a crash.

**Comment by alainjobart** [comment]
Do you have the stack trace of the panic? That would be helpful!

**Comment by alainjobart** [comment]
Ignore my previous comment, the check Sugu added is at the wrong place. If the policy is WriteGlobalBuffer, currentEphemeralPacket would be nil.

File: go/mysql/conn.go [hunk 1]
```diff
@@ -548,17 +551,29 @@ func (c *Conn) startEphemeralPacket(length int) []byte {
...
+	if w == nil {
+		panic("w (io.Writer) is nil")
+	}
 	if direct {
 		w = c.conn
+		if w == nil {
+			panic("w (c.conn) is nil")
+		}
+	}
+	if c.currentEphemeralBuffer == nil {
```
Comment (alainjobart) [inline go/mysql/conn.go]: this check is at the wrong place, ephemeralWriteGlobalBuffer policy doesn't use this buffer.

**Comment by sougou** [comment]
Sorry. That was my bad on the ephemeral check. I should have moved under under the case statements. But the warning will do for now.

**Comment by slanning** [comment]
I'm still replaying queries, but "unfortunately" it hasn't crashed yet today.

**Comment by sougou** [comment]
Closing this, as we now know the root cause of the vtexplain problem.

**Pull request closed without merging**